Add the test script to trigger and capture cpuidle state (New)

seankingyang opened the pull request:
## Description
- Add the trigger and capture CPU cpuidle state
- For ARM platform also have the CPU idle feature, and this script is similar fwts cstates with more useful info.
- This script also can be run on X86 platform.

## Resolved issues
Miss coverage of the CPU idle.


## Documentation



## Tests

ARM w/o CPUidle driver (this is a true issue):
```
root@ubuntu:/home/ubuntu/CPU# ./cpu_idle_state.py 
[INFO] Starting CPU idle state test
[INFO] This test checks if all processors have the same number of idle states and if idle state transitions happen during alternating load patterns
[INFO] Found 2 CPUs in system
[ERROR] ERROR: No CPUs with cpuidle support found
root@ubuntu:/home/ubuntu/CPU# echo $?
1
```

ARM w/ CPUidle driver (some state can't entry during the trigger idle state testing but the usage of state w/ existing usage so mark as pass )
```
root@mtk-genio:/home/ubuntu/CPU# ./cpu_idle_state.py 
[INFO] Starting CPU idle state test
[INFO] This test checks if all processors have the same number of idle states and if idle state transitions happen during alternating load patterns
[INFO] Found 8 CPUs in system
[INFO] CPU 0: Found 3 idle states
[INFO] CPU 1: Found 3 idle states
[INFO] CPU 2: Found 3 idle states
[INFO] CPU 3: Found 3 idle states
[INFO] CPU 4: Found 3 idle states
[INFO] CPU 5: Found 3 idle states
[INFO] CPU 6: Found 3 idle states
[INFO] CPU 7: Found 3 idle states
[INFO] Testing 8 CPUs with cpuidle support
[INFO] Testing CPU 0 (1/8)
[INFO] CPU 0: State 2 (clusteroff_l) was used
[INFO] CPU 0: State 0 (WFI) was used
[INFO] CPU 0: State 1 (cpuoff_l) was used
[INFO] CPU 0: All idle states reached, stopping early
[INFO] CPU 0: SUCCESS - All states reached: cpuoff_l, clusteroff_l, WFI
[INFO] Testing CPU 1 (2/8)
[INFO] CPU 1: State 2 (clusteroff_l) was used
[INFO] CPU 1: State 0 (WFI) was used
[INFO] CPU 1: NOTE - States with existing usage (not tested): cpuoff_l
[INFO] Testing CPU 2 (3/8)
[INFO] CPU 2: State 1 (cpuoff_l) was used
[INFO] CPU 2: State 2 (clusteroff_l) was used
[INFO] CPU 2: State 0 (WFI) was used
[INFO] CPU 2: All idle states reached, stopping early
[INFO] CPU 2: SUCCESS - All states reached: cpuoff_l, clusteroff_l, WFI
[INFO] Testing CPU 3 (4/8)
[INFO] CPU 3: State 2 (clusteroff_l) was used
[INFO] CPU 3: State 0 (WFI) was used
[INFO] CPU 3: NOTE - States with existing usage (not tested): cpuoff_l
[INFO] Testing CPU 4 (5/8)
[INFO] CPU 4: State 2 (clusteroff_b) was used
[INFO] CPU 4: State 1 (cpuoff_b) was used
[INFO] CPU 4: State 0 (WFI) was used
[INFO] CPU 4: All idle states reached, stopping early
[INFO] CPU 4: SUCCESS - All states reached: cpuoff_b, clusteroff_b, WFI
[INFO] Testing CPU 5 (6/8)
[INFO] CPU 5: State 2 (clusteroff_b) was used
[INFO] CPU 5: State 1 (cpuoff_b) was used
[INFO] CPU 5: NOTE - States with existing usage (not tested): WFI
[INFO] Testing CPU 6 (7/8)
[INFO] CPU 6: State 2 (clusteroff_b) was used
[INFO] CPU 6: State 1 (cpuoff_b) was used
[INFO] CPU 6: State 0 (WFI) was used
[INFO] CPU 6: All idle states reached, stopping early
[INFO] CPU 6: SUCCESS - All states reached: cpuoff_b, clusteroff_b, WFI
[INFO] Testing CPU 7 (8/8)
[INFO] CPU 7: State 1 (cpuoff_b) was used
[INFO] CPU 7: State 2 (clusteroff_b) was used
[INFO] CPU 7: State 0 (WFI) was used
[INFO] CPU 7: All idle states reached, stopping early
[INFO] CPU 7: SUCCESS - All states reached: cpuoff_b, clusteroff_b, WFI
[INFO] 
Test completed: 8/8 CPUs passed
[INFO] SUCCESS: All CPUs reached all their idle states
root@mtk-genio:/home/ubuntu/CPU# echo $?
0
```

seankingyang committed:
Add the test script to trigger and capture cpuidle state

seankingyang committed:
modify the mock call to fit the py3.5

rickwu666666 reviewed (commented):
I've tried the script on two of ARM devices. And seems it catch one of device didn't support cpuidle. 
```
[INFO] Starting CPU idle state test
[INFO] This test checks if all processors have the same number of idle states and if idle state transitions happen during alternating load patterns
[INFO] Found 2 CPUs in system
[ERROR] ERROR: No CPUs with cpuidle support found
```
I think this test is valuable. But what the reason not implement the test job for it?

stanley31huang commented on providers/base/bin/cpu_idle_state.py:
```diff
@@ -0,0 +1,583 @@
+        b = 3.121213
+
+        for _ in range(100):
+            a = a * b
+            b = a * a
+            # Ensure a is positive before taking square root
+            a = a - b + math.sqrt(abs(a))
+            a = a * b
+            b = a * a
+            a = a - b + math.sqrt(abs(a))
+            a = a * b
+            b = a * a
+            a = a - b + math.sqrt(abs(a))
+            a = a * b
+            b = a * a
+            a = a - b + math.sqrt(abs(a))
+
+    def cpu_benchmark(self, cpu_id: int) -> float:
+        """Run CPU benchmark on specific CPU, similar to FWTS
+        fwts_cpu_benchmark()
+        It will burn CPU cycles for ~250ms.
+        """
+        # Set CPU affinity
+        old_affinity = self.set_cpu_affinity(cpu_id)
+
+        try:
+            start_time = time.time()
+            loops = 0
+
+            # Run benchmark for ~250ms (same as FWTS)
+            while (time.time() - start_time) < 0.25:
+                self.burn_cpu_cycles()
+                loops += 1
+
+            duration = time.time() - start_time
+            return loops / duration if duration > 0 else 0
+
+        finally:
+            # Restore CPU affinity
+            self.restore_cpu_affinity(old_affinity)
```
I recommend we keep these functions as simple as possible. A key advantage of this approach is that it will be clear what CPU affinity settings we are attempting to restore if any issues arise.
```suggestion
    def set_cpu_affinity(self, cpu_id: int) -> List[int]:
        """Set CPU affinity to a specific CPU and return previous affinity"""
        try:
            psutil.Process().cpu_affinity([cpu_id])
        except psutil.AccessDenied:
            msg = "Cannot set CPU affinity to CPU {} ".format(cpu_id)
            msg += "- need root privileges"
            raise CpuAffinityError(msg)

    def burn_cpu_cycles(self):
        """
        Burn CPU cycles similar to FWTS fwts_cpu_burn_cycles().
        It use the floating point operations to burn CPU cycles.
        """
        a = 1.234567
        b = 3.121213

        for _ in range(100):
            a = a * b
            b = a * a
            # Ensure a is positive before taking square root
            a = a - b + math.sqrt(abs(a))
            a = a * b
            b = a * a
            a = a - b + math.sqrt(abs(a))
            a = a * b
            b = a * a
            a = a - b + math.sqrt(abs(a))
            a = a * b
            b = a * a
            a = a - b + math.sqrt(abs(a))

    def cpu_benchmark(self, cpu_id: int) -> float:
        """Run CPU benchmark on specific CPU, similar to FWTS
        fwts_cpu_benchmark()
        It will burn CPU cycles for ~250ms.
        """
        # Set CPU affinity
        old_affinity = self.get_cpu_afinity()
        self.set_cpu_affinity(cpu_id)

        try:
            start_time = time.time()
            loops = 0

            # Run benchmark for ~250ms (same as FWTS)
            while (time.time() - start_time) < 0.25:
                self.burn_cpu_cycles()
                loops += 1

            duration = time.time() - start_time
            return loops / duration if duration > 0 else 0

        finally:
            # Restore CPU affinity
            self.set_cpu_affinity(old_affinity)
            self.logger("restore CPU affinity failed")
```

seankingyang committed:
Apply Stanley's suggestion: combine the set and restore the cpu affinity and seperate the get affinity from the set function

stanley31huang reviewed (dismissed):
LGTM

seankingyang committed:
correct the unit test case

seankingyang committed:
fix the assert_called_once can't be use in py3.5

seankingyang committed:
Apply Copilot review suggestions on the unit tests

- test_burn_cpu_cycles_loop_execution: count math.sqrt calls (400 per
  run) instead of asserting a 1s wall-clock bound, which is flaky on
  slow or heavily loaded hosts
- test_main_root_check_logic: actually invoke main() as non-root and
  assert it exits 1 with the root-required error message

rickwu666666 reviewed (approved):
LGTM
